### PR TITLE
Feat: Refactor dipole fitting pytorch

### PR DESCRIPTION
**Pull request by anyangml**
This PR is to provide implementation of equivariant diplole fitting in pytorch and backend-independent numpy.

### CODE REVIEW

File: source/tests/pt/model/test_dipole_fitting.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
+# SPDX-License-Identifier: LGPL-3.0-or-later
+import itertools
+import unittest
+
+import numpy as np
+import torch
+
+from deepmd.dpmodel.fitting import DipoleFitting as DPDipoleFitting
```
Comment (anyangml) [inline source/tests/pt/model/test_dipole_fitting.py:8]: I have not tested consistency with TF yet. Is there any available fixture we can use to test cross framework consistency? probably need to export TF model weights and import to PT model.

Comment (njzjz) [reply source/tests/pt/model/test_dipole_fitting.py:8]: See #3263

File: source/tests/pt/model/test_dipole_fitting.py
Comment (njzjz) [inline source/tests/pt/model/test_dipole_fitting.py:1]: Need to add a test for `DeepDipole`.

File: deepmd/dpmodel/fitting/dipole_fitting.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+from .general_fitting import (
+    GeneralFitting,
+)
+
+
+@fitting_check_output
+class DipoleFitting(GeneralFitting):
+    r"""Fitting the energy (or a rotationally invariant porperty of `dim_out`) of the system. The force and the virial can also be trained.
+
+    Lets take the energy fitting task as an example.
```
Comment (njzjz) [inline deepmd/dpmodel/fitting/dipole_fitting.py:27]: "Energy" appears in `DipoleFitting`, which looks wrong.

File: deepmd/dpmodel/fitting/dipole_fitting.py [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+)
+
+from .general_fitting import (
+    GeneralFitting,
+)
+
+
+@fitting_check_output
+class DipoleFitting(GeneralFitting):
+    r"""Fitting the energy (or a rotationally invariant porperty of `dim_out`) of the system. The force and the virial can also be trained.
```
Comment (njzjz) [inline deepmd/dpmodel/fitting/dipole_fitting.py:25]: Isn't "energy" dipole?

File: deepmd/dpmodel/fitting/dipole_fitting.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+    ):
+        # seed, uniform_seed are not included
+        if tot_ener_zero:
+            raise NotImplementedError("tot_ener_zero is not implemented")
+        if spin is not None:
+            raise NotImplementedError("spin is not implemented")
+        if use_aparam_as_mask:
+            raise NotImplementedError("use_aparam_as_mask is not implemented")
+        if use_aparam_as_mask:
+            raise NotImplementedError("use_aparam_as_mask is not implemented")
```
[the remark this answers is not in the document]
Comment (njzjz) [reply deepmd/dpmodel/fitting/dipole_fitting.py:135]: Duplicated as line 132

File: deepmd/pt/model/task/fitting.py [hunk 5]
```diff
@@ -421,11 +421,12 @@ def __init__(
 
         self.emask = AtomExcludeMask(self.ntypes, self.exclude_types)
 
+        net_dim_out = self._net_out_dim()
         # init constants
         if bias_atom_e is None:
-            bias_atom_e = np.zeros([self.ntypes, self.dim_out])
+            bias_atom_e = np.zeros([self.ntypes, net_dim_out])
```
Comment (njzjz) [inline deepmd/pt/model/task/fitting.py:427]: dtype should be set

File: deepmd/dpmodel/fitting/dipole_fitting.py [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+        dim_out: int,
+        dim_rot_mat: int,
+        neuron: List[int] = [120, 120, 120],
+        resnet_dt: bool = True,
+        numb_fparam: int = 0,
+        numb_aparam: int = 0,
+        rcond: Optional[float] = None,
+        tot_ener_zero: bool = False,
+        trainable: Optional[List[bool]] = None,
+        atom_ener: Optional[List[float]] = None,
```
Comment (njzjz) [inline deepmd/dpmodel/fitting/dipole_fitting.py:117]: should be `Optional[List[Optional[float]]]`

File: deepmd/pt/model/task/dipole.py
Comment (njzjz) [inline deepmd/pt/model/task/dipole.py:1]: Should a test of consistency be added in this PR, like #3286?

**Review by njzjz — CHANGES_REQUESTED** [review]

File: deepmd/pt/model/task/dipole.py
Comment (anyangml) [reply deepmd/pt/model/task/dipole.py:1]: not in this PR. will merge this PR, and a polar fitting PR, then after 3286 is merged, we can add consistency test.

File: deepmd/pt/model/task/dipole.py [hunk 7]
```diff
@@ -1,67 +1,135 @@
...
 
-        filter_layers = []
-        one = ResidualDeep(
-            0, embedding_width, neuron, 0.0, out_dim=self.out_dim, resnet_dt=resnet_dt
+    def __init__(
+        self,
+        var_name: str,
+        ntypes: int,
+        dim_descrpt: int,
+        dim_out: int,
```
Comment (wanghan-iapcm) [inline deepmd/pt/model/task/dipole.py:63]: should not be a parameter, because the output dimension is always 3.

File: deepmd/pt/model/task/dipole.py [hunk 8]
```diff
@@ -1,67 +1,135 @@
...
+        rcond: Optional[float] = None,
+        seed: Optional[int] = None,
+        **kwargs,
+    ):
+        self.dim_rot_mat = dim_rot_mat
+        super().__init__(
+            var_name=var_name,
+            ntypes=ntypes,
+            dim_descrpt=dim_descrpt,
+            dim_out=dim_out,
```
Comment (wanghan-iapcm) [inline deepmd/pt/model/task/dipole.py:81]: I find the `GeneralFitting` should not accept the dim_out parameter, because the network output dim is left as an abstract method that is implemented by the derived class.

File: deepmd/pt/model/task/dipole.py [hunk 9]
```diff
@@ -1,67 +1,135 @@
...
+        neuron: List[int] = [128, 128, 128],
+        resnet_dt: bool = True,
+        numb_fparam: int = 0,
+        numb_aparam: int = 0,
+        activation_function: str = "tanh",
+        precision: str = DEFAULT_PRECISION,
+        distinguish_types: bool = False,
+        rcond: Optional[float] = None,
+        seed: Optional[int] = None,
+        **kwargs,
```
Comment (wanghan-iapcm) [inline deepmd/pt/model/task/dipole.py:74]: please support the parameter `exclude_types`

File: source/tests/pt/model/test_dipole_fitting.py [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+            [True, False],
+            [0, 3],
+            [0, 4],
+        ):
+            ft0 = DipoleFittingNet(
+                "foo",
+                self.nt,
+                dd0.dim_out,
+                3,
+                dim_rot_mat=100,
```
Comment (wanghan-iapcm) [inline source/tests/pt/model/test_dipole_fitting.py:62]:
```suggestion
                dim_rot_mat=dd0.get_dim_emb(),
```
The same suggestion for other `dim_rot_mat`

**Review by wanghan-iapcm — CHANGES_REQUESTED** [review]

File: source/tests/pt/model/test_dipole_fitting.py
Comment (anyangml) [reply source/tests/pt/model/test_dipole_fitting.py:1]: will add in future PR.

**Comment by anyangml** [comment]
@njzjz can we merge this PR before https://github.com/deepmodeling/deepmd-kit/pull/3286, and then merge the PR for polar.

**Review by wanghan-iapcm — APPROVED** [review]

File: deepmd/pt/model/task/dipole.py [hunk 11]
```diff
@@ -1,67 +1,152 @@
...
 
-        filter_layers = []
-        one = ResidualDeep(
-            0, embedding_width, neuron, 0.0, out_dim=self.out_dim, resnet_dt=resnet_dt
+    def __init__(
+        self,
+        var_name: str,
+        ntypes: int,
+        dim_descrpt: int,
+        dim_rot_mat: int,
```
Comment (wanghan-iapcm) [inline deepmd/pt/model/task/dipole.py:67]: please not my comment in #3286 , the `embedding_width` is preferred.

Comment (anyangml) [reply deepmd/pt/model/task/dipole.py:67]: if it's not fixed in #3286, I will fix it in the polar fitting pr.

**Review by njzjz — APPROVED** [review]

**Review by iProzd — APPROVED** [review]

File: deepmd/pt/model/task/dipole.py [hunk 12]
```diff
@@ -1,67 +1,152 @@
...
 log = logging.getLogger(__name__)
 
 
-class DipoleFittingNet(Fitting):
-    def __init__(
-        self, ntypes, embedding_width, neuron, out_dim, resnet_dt=True, **kwargs
-    ):
-        """Construct a fitting net for dipole.
+class DipoleFittingNet(GeneralFitting):
+    """Construct a general fitting net.
```
[the remark this answers is not in the document]
Comment (anyangml) [reply deepmd/pt/model/task/dipole.py:28]: will fix in the next PR.
